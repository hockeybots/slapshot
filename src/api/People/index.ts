import idx from 'idx';

import { PEOPLE_ENDPOINT } from '../../endpoints';
import { Player } from '../types';

import Endpoint from '../Endpoint';

/**
 * People endpoint wrapper
 * @extends Endpoint
 * @description An abstraction of the {PEOPLE_ENDPOINT} with a fluent API
 */
class People extends Endpoint {
  /**
   * This method will transform API data in to a Player object.
   * @param apiData {object} - The conference object of the response from the NHL API
   * @returns {Player}
   */
  public static async toPlayer(apiData: any): Promise<Player> {
    const player: Player = {
      age: apiData.currentAge,
      birthDate: new Date(apiData.birthDate),
      birthplace: `${apiData.birthCity}, ${apiData.birthStateProvince || ''} ${apiData.birthCountry}`,
      dominantSide: apiData.shootsCatches,
      firstName: apiData.firstName,
      fullName: apiData.fullName,
      height: apiData.height,
      id: apiData.id,
      isAlternateCaptain: apiData.alternateCaptain,
      isCaptain: apiData.captain,
      isRookie: apiData.rookie,
      jerseyNumber: apiData.primaryNumber,
      lastName: apiData.lastName,
      position: apiData.primaryPosition.name,
      rosterStatus: apiData.rosterStatus,
      weight: apiData.weight,
    };
    return player;
  }

  private playerIds: Array<number>;
  constructor(...ids: Array<number>) {
    super();
    this.playerIds = ids;
    this.initUri();
  }
  public clear() {
    this.initUri();
  }
  public initUri() {
    this.uri = `${PEOPLE_ENDPOINT}?personIds=${this.playerIds.join(',')}`;
  }
  /**
   * @description This method is used after building of the URI is complete. It will fetch and parse the NHL API data.
   * @returns {Player[]}
   */
  public async data(): Promise<Array<Player>> {
    try {
      const apiData = await this.load();
      return this.parseData<Player>(apiData, 'people', People.toPlayer);
    } catch (error) {
      return Promise.reject(error);
    }
  }
}

export default People;
