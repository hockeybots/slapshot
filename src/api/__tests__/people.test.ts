import People from '../People';

describe('Api', () => {
  describe('People', () => {
    describe('toPlayer', () => {
      const mockApiData = {
        currentAge: 22,
        birthDate: '1990-01-01',
        birthCity: 'City of Angels',
        birthStateProvince: 'CA',
        birthCountry: 'USA',
        shootsCatches: 'R',
        firstName: 'Cool',
        fullName: 'Cool Guy',
        height: '6\' 5"',
        id: 1,
        alternateCaptain: false,
        captain: true,
        rookie: false,
        primaryNumber: '99',
        lastName: 'Guy',
        primaryPosition: {
          name: 'Goalie',
        },
        rosterStatus: 'A',
        weight: 222,
      };
      it('should return a player object with the correct data', async () => {
        const player = await People.toPlayer(mockApiData);
        expect(player).toMatchSnapshot();
      });
    });
  });
});
