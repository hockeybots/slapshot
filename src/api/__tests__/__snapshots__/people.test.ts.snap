// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api People toPlayer should return a player object with the correct data 1`] = `
Object {
  "age": 22,
  "birthDate": 1990-01-01T00:00:00.000Z,
  "birthplace": "City of Angels, CA USA",
  "dominantSide": "R",
  "firstName": "Cool",
  "fullName": "Cool Guy",
  "height": "6' 5\\"",
  "id": 1,
  "isAlternateCaptain": false,
  "isCaptain": true,
  "isRookie": false,
  "jerseyNumber": "99",
  "lastName": "Guy",
  "position": "Goalie",
  "rosterStatus": "A",
  "weight": 222,
}
`;
